# Using SQL to analyse the statistics about the countries.

For this project, we will work the data from the [CIA World Factbook](https://www.cia.gov/library/publications/the-world-factbook/), which has the statistics about all the countries on Earth. 

The dataset contains demographic information like:
- `population` : The population as of 2015.
- `population_growth` : The annual population growth rate, as a percentage.
- `area` : The total land and water area.

## 1. Loding the data
First, lets connect the Notebook to the database file which is a sql file.

In [5]:
%%capture
%load_ext sql
%sql sqlite:///factbook.db

'Connected: None@factbook.db'

In [10]:
%%sql
SELECT *
    FROM sqlite_master
    WHERE type = 'table';

Done.


[('table', 'sqlite_sequence', 'sqlite_sequence', 3, 'CREATE TABLE sqlite_sequence(name,seq)'),
 ('table', 'facts', 'facts', 47, 'CREATE TABLE "facts" ("id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, "code" varchar(255) NOT NULL, "name" varchar(255) NOT NULL, "area" integer, "a ... (4 characters truncated) ... land" integer, "area_water" integer, "population" integer, "population_growth" float, "birth_rate" float, "death_rate" float, "migration_rate" float)')]

The above output clearly states that there is a table with name `facts` which has columns like `area_land`, `area_water`, `population` etc 

In [13]:
%%sql
SELECT *
    FROM facts
    LIMIT 5;

Done.


[(1, 'af', 'Afghanistan', 652230, 652230, 0, 32564342, 2.32, 38.57, 13.89, 1.51),
 (2, 'al', 'Albania', 28748, 27398, 1350, 3029278, 0.3, 12.92, 6.58, 3.3),
 (3, 'ag', 'Algeria', 2381741, 2381741, 0, 39542166, 1.84, 23.67, 4.31, 0.92),
 (4, 'an', 'Andorra', 468, 468, 0, 85580, 0.12, 8.13, 6.96, 0.0),
 (5, 'ao', 'Angola', 1246700, 1246700, 0, 19625353, 2.78, 38.78, 11.49, 0.46)]

## Quering min and max of population and population growth

In [20]:
%%sql
SELECT MIN(population), MAX(population), MIN(population_growth), MAX(population_growth)
    FROM facts;

Done.


[(0, 7256490011, 0.0, 4.02)]

One can infer the following
- There is a country with `0` population
- Max population is `7.2 Billion`
- Minimum population growth is `0`
- Max population greoth is `4.02`

The last two points are probable, but the first two doesn't seem true. Lets explore these rows in the next cell.

## Exploring outliers in the population

In [21]:
%%sql
SELECT *
    FROM facts
    WHERE population = (SELECT MIN(population) FROM FACTS);

Done.


[(250, 'ay', 'Antarctica', None, 280000, None, 0, None, None, None, None)]

**The country with 0 population is Antartica which looks correct.**

In [23]:
%%sql
SELECT *
    FROM facts
    WHERE population = (SELECT MAX(population) FROM facts);

Done.


[(261, 'xx', 'World', None, None, None, 7256490011, 1.08, 18.6, 7.8, None)]

**Lol, the row which has `7.2 Billion` population is not a country but that row belongs to the overall `World`.**

So neglecting world and quering again results in 

In [26]:
%%sql
SELECT MIN(population), MAX(population), MIN(population_growth), MAX(population_growth)
    FROM facts
    WHERE name <> 'World';

Done.


[(0, 1367485388, 0.0, 4.02)]

The max population returned now is 1.36 Billion. Looks like China, we will confirm this in the following cell

In [28]:
%%sql
SELECT *
    FROM facts
    WHERE population = (SELECT MAX(population) FROM facts WHERE name <> 'World');

Done.


[(37, 'ch', 'China', 9596960, 9326410, 270550, 1367485388, 0.45, 12.49, 7.53, 0.44)]

## Exploring average value of population and area among the countries

In [29]:
%%sql
SELECT AVG(population), AVG(area)
    FROM facts
    WHERE name <> 'World';

Done.


[(32242666.56846473, 555093.546184739)]

It can be seen that average population is 32 million and average area is 555,000 sq Km^2.

## Finding Densely populated countries

To find densely populated countries we look at the countries which have
- `population` higher than the average population, and,
- `area` below the average

Ofcourse this is not ideal, because countries like 
- `India` may be left out this. These may not meet the second criteria. 
- City states like `Singapore` etc may also be left out. Because they may not have as high population as 32 million in a city.

In [30]:
%%sql
SELECT *
    FROM facts
    WHERE (population > (SELECT AVG(population)
    FROM facts
    WHERE name <> 'World')) AND (area < (SELECT AVG(area)
    FROM facts
    WHERE name <> 'World'));

Done.


[(14, 'bg', 'Bangladesh', 148460, 130170, 18290, 168957745, 1.6, 21.14, 5.61, 0.46),
 (65, 'gm', 'Germany', 357022, 348672, 8350, 80854408, 0.17, 8.47, 11.42, 1.24),
 (80, 'iz', 'Iraq', 438317, 437367, 950, 37056169, 2.93, 31.45, 3.77, 1.62),
 (83, 'it', 'Italy', 301340, 294140, 7200, 61855120, 0.27, 8.74, 10.19, 4.1),
 (85, 'ja', 'Japan', 377915, 364485, 13430, 126919659, 0.16, 7.93, 9.51, 0.0),
 (91, 'ks', 'Korea, South', 99720, 96920, 2800, 49115196, 0.14, 8.19, 6.75, 0.0),
 (120, 'mo', 'Morocco', 446550, 446300, 250, 33322699, 1.0, 18.2, 4.81, 3.36),
 (138, 'rp', 'Philippines', 300000, 298170, 1830, 100998376, 1.61, 24.27, 6.11, 2.09),
 (139, 'pl', 'Poland', 312685, 304255, 8430, 38562189, 0.09, 9.74, 10.19, 0.46),
 (163, 'sp', 'Spain', 505370, 498980, 6390, 48146134, 0.89, 9.64, 9.04, 8.31),
 (173, 'th', 'Thailand', 513120, 510890, 2230, 67976405, 0.34, 11.19, 7.8, 0.0),
 (182, 'ug', 'Uganda', 241038, 197100, 43938, 37101745, 3.24, 43.79, 10.69, 0.74),
 (185, 'uk', 'United Kingdom', 243610, 241930, 1680, 64088222, 0.54, 12.17, 9.35, 2.54),
 (192, 'vm', 'Vietnam', 331210, 310070, 21140, 94348835, 0.97, 15.96, 5.93, 0.3)]

As expected, countries like `Singapore`, `Malta`, `India` and `China` don't feature in this list. So we will change the evalution metric
## Changing the density evaluation metric
Instead of taking above average population and below average area, we will add another column density and list the countries above the average density. This will be more representative.

In [32]:
%%sql
SELECT *, population / CAST(area AS Float) AS density
    FROM facts
    WHERE (name <> 'World') AND (density > (SELECT AVG(population / CAST(area AS Float)) FROM facts WHERE name <> 'World') );

Done.


[(13, 'ba', 'Bahrain', 760, 760, 0, 1346613, 2.41, 13.66, 2.69, 13.09, 1771.8592105263158),
 (14, 'bg', 'Bangladesh', 148460, 130170, 18290, 168957745, 1.6, 21.14, 5.61, 0.46, 1138.0691432035565),
 (15, 'bb', 'Barbados', 430, 430, 0, 290604, 0.31, 11.87, 8.44, 0.3, 675.8232558139534),
 (91, 'ks', 'Korea, South', 99720, 96920, 2800, 49115196, 0.14, 8.19, 6.75, 0.0, 492.53104693140796),
 (97, 'le', 'Lebanon', 10400, 10230, 170, 6184701, 0.86, 14.59, 4.88, 1.1, 594.6827884615385),
 (108, 'mv', 'Maldives', 298, 298, 0, 393253, 0.08, 15.75, 3.89, 12.68, 1319.6409395973155),
 (110, 'mt', 'Malta', 316, 316, 0, 413965, 0.31, 10.18, 9.09, 1.98, 1310.01582278481),
 (113, 'mp', 'Mauritius', 2040, 2030, 10, 1339827, 0.64, 13.29, 6.91, 0.0, 656.7779411764706),
 (117, 'mn', 'Monaco', 2, 2, 0, 30535, 0.12, 6.65, 9.24, 3.83, 15267.5),
 (123, 'nr', 'Nauru', 21, 21, 0, 9540, 0.55, 24.95, 5.87, 13.63, 454.2857142857143),
 (144, 'rw', 'Rwanda', 26338, 24668, 1670, 12661733, 2.56, 33.75, 8.96, 0.85, 480.74010934771053),
 (149, 'sm', 'San Marino', 61, 61, 0, 33020, 0.82, 8.63, 8.45, 8.03, 541.311475409836),
 (156, 'sn', 'Singapore', 697, 687, 10, 5674472, 1.89, 8.27, 3.43, 14.05, 8141.279770444763),
 (196, 'tw', 'Taiwan', 35980, 32260, 3720, 23415126, 0.23, 8.47, 7.11, 0.89, 650.7817120622568),
 (204, 'hk', 'Hong Kong', 1108, 1073, 35, 7141106, 0.38, 9.23, 7.07, 1.68, 6445.041516245487),
 (205, 'mc', 'Macau', 28, 28, 0, 592731, 0.8, 8.88, 4.22, 3.37, 21168.964285714286),
 (213, 'rn', 'Saint Martin', 54, 54, None, 31754, None, None, None, None, 588.0370370370371),
 (216, 'aa', 'Aruba', 180, 180, 0, 112162, 1.33, 12.56, 8.18, 8.92, 623.1222222222223),
 (218, 'sk', 'Sint Maarten', 34, 34, 0, 39689, 1.51, 13.0, 4.51, 6.63, 1167.3235294117646),
 (227, 'bd', 'Bermuda', 54, 54, 0, 70196, 0.5, 11.33, 8.23, 1.88, 1299.9259259259259),
 (233, 'gi', 'Gibraltar', 6, 6, 0, 29258, 0.24, 14.08, 8.37, 3.28, 4876.333333333333),
 (234, 'gk', 'Guernsey', 78, 78, 0, 66080, 0.34, 9.84, 8.78, 2.32, 847.1794871794872),
 (235, 'je', 'Jersey', 116, 116, 0, 97294, 0.8, 11.91, 7.68, 3.76, 838.7413793103449),
 (251, 'gz', 'Gaza Strip', 360, 360, 0, 1869055, 2.81, 31.11, 3.04, 0.0, 5191.819444444444),
 (254, 'we', 'West Bank', 5860, 5640, 220, 2785366, 1.95, 22.99, 3.5, 0.0, 475.3184300341297)]

It is interesting that countries like `India` don't make up this list. So it is essential to manually check what is the average density of the world and of countries like `India`

In [34]:
%%sql
SELECT AVG(population / CAST(area AS Float)) AS avg_density FROM facts WHERE name <> 'World' ;

Done.


[(419.66252469247945,)]

In [37]:
%%sql
SELECT name, (population / CAST(area AS Float)) AS density FROM facts WHERE name in ('India', 'China','Indonesia') ;

Done.


[('China', 142.4915168970174),
 ('India', 380.7713541630226),
 ('Indonesia', 134.41029125224657)]

**It can be observed that the query outputed is correct and the density of these countries are less than the average**

## Which are the unsafe countries
It would be interesting to see the countries with higher `death rates`

In [59]:
%%sql
SELECT name, birth_rate, death_rate, migration_rate
    FROM facts
    WHERE name <> 'World'
    ORDER BY death_rate DESC;

Done.


[('Lesotho', 25.47, 14.89, 7.36),
 ('Ukraine', 10.72, 14.46, 2.25),
 ('Bulgaria', 8.92, 14.44, 0.29),
 ('Guinea-Bissau', 33.38, 14.33, 0.0),
 ('Latvia', 10.0, 14.31, 6.26),
 ('Chad', 36.6, 14.28, 3.45),
 ('Lithuania', 10.1, 14.27, 6.27),
 ('Namibia', 19.8, 13.91, 0.0),
 ('Afghanistan', 38.57, 13.89, 1.51),
 ('Central African Republic', 35.08, 13.8, 0.0),
 ('Russia', 11.6, 13.69, 1.69),
 ('Serbia', 9.08, 13.66, 0.0),
 ('Somalia', 40.45, 13.62, 8.49),
 ('Swaziland', 24.67, 13.56, 0.0),
 ('Botswana', 20.96, 13.39, 4.56),
 ('Belarus', 10.7, 13.36, 0.7),
 ('Gabon', 34.49, 13.12, 2.03),
 ('Nigeria', 37.64, 12.9, 0.22),
 ('Mali', 44.99, 12.89, 2.26),
 ('Hungary', 9.16, 12.73, 1.33),
 ('Zambia', 42.13, 12.67, 0.68),
 ('Moldova', 12.0, 12.59, 9.67),
 ('Niger', 45.45, 12.42, 0.56),
 ('Estonia', 10.51, 12.4, 3.6),
 ('Croatia', 9.45, 12.18, 1.39),
 ('Mozambique', 38.58, 12.1, 1.98),
 ('Romania', 9.14, 11.9, 0.24),
 ('Burkina Faso', 42.03, 11.72, 0.0),
 ('Angola', 38.78, 11.49, 0.46),
 ('Germany', 8.47, 11.42, 1.24),
 ('Slovenia', 8.42, 11.37, 0.37),
 ('Greece', 8.66, 11.09, 2.32),
 ('Portugal', 9.27, 11.02, 2.67),
 ('Georgia', 12.74, 10.82, 2.7),
 ('Sierra Leone', 37.03, 10.81, 2.77),
 ('Uganda', 43.79, 10.69, 0.74),
 ('Czech Republic', 9.63, 10.34, 2.33),
 ('Denmark', 10.27, 10.25, 2.2),
 ('European Union', 10.2, 10.2, 2.5),
 ('Italy', 8.74, 10.19, 4.1),
 ('Poland', 9.74, 10.19, 0.46),
 ('Zimbabwe', 32.26, 10.13, 0.0),
 ('Cameroon', 36.17, 10.11, 0.15),
 ('Congo, Democratic Republic of the', 34.88, 10.07, 0.27),
 ('Isle of Man', 11.1, 10.06, 6.56),
 ('Congo, Republic of the', 35.85, 10.0, 5.9),
 ('South Africa', 20.75, 9.91, 2.42),
 ('Finland', 10.72, 9.83, 3.1),
 ('Bosnia and Herzegovina', 8.87, 9.75, 0.38),
 ('Slovakia', 9.91, 9.74, 0.04),
 ('Saint Pierre and Miquelon', 7.42, 9.72, 8.49),
 ('Liberia', 34.41, 9.69, 0.0),
 ('Belgium', 11.41, 9.63, 5.87),
 ("Cote d'Ivoire", 28.67, 9.55, 0.0),
 ('Japan', 7.93, 9.51, 0.0),
 ('Guinea', 35.74, 9.46, 0.0),
 ('Uruguay', 13.07, 9.45, 0.9),
 ('Montenegro', 10.42, 9.43, None),
 ('Austria', 9.41, 9.42, 5.56),
 ('Sweden', 11.99, 9.4, 5.42),
 ('United Kingdom', 12.17, 9.35, 2.54),
 ('Armenia', 13.61, 9.34, 5.8),
 ('Burundi', 42.01, 9.27, 0.0),
 ('Monaco', 6.65, 9.24, 3.83),
 ('Korea, North', 14.52, 9.21, 0.04),
 ('France', 12.38, 9.16, 1.09),
 ('Malta', 10.18, 9.09, 1.98),
 ('Macedonia', 11.55, 9.08, 0.48),
 ('Spain', 9.64, 9.04, 8.31),
 ('Rwanda', 33.75, 8.96, 0.85),
 ('Guernsey', 9.84, 8.78, 2.32),
 ('Tuvalu', 23.74, 8.74, 6.81),
 ('Faroe Islands', 13.77, 8.71, 0.0),
 ('Puerto Rico', 10.86, 8.67, 8.15),
 ('Netherlands', 10.83, 8.66, 1.95),
 ('Trinidad and Tobago', 13.46, 8.56, 6.25),
 ('Virgin Islands', 10.31, 8.54, 7.67),
 ('Greenland', 14.48, 8.49, 5.98),
 ('Senegal', 34.52, 8.46, 1.59),
 ('San Marino', 8.63, 8.45, 8.03),
 ('Barbados', 11.87, 8.44, 0.3),
 ('Canada', 10.28, 8.42, 5.66),
 ('Malawi', 41.56, 8.41, 0.0),
 ('Gibraltar', 14.08, 8.37, 3.28),
 ('Western Sahara', 30.24, 8.34, None),
 ('Bermuda', 11.33, 8.23, 1.88),
 ('Benin', 36.02, 8.21, 0.0),
 ('Kazakhstan', 19.15, 8.21, 0.41),
 ('Mauritania', 31.34, 8.2, 0.83),
 ('Curacao', 13.8, 8.2, 1.4),
 ('Equatorial Guinea', 33.31, 8.19, 0.0),
 ('Ethiopia', 37.27, 8.19, 0.22),
 ('South Sudan', 36.91, 8.18, 11.47),
 ('Aruba', 12.56, 8.18, 8.92),
 ('United States', 12.49, 8.15, 3.86),
 ('Switzerland', 10.5, 8.13, 4.74),
 ('Norway', 12.14, 8.12, 7.25),
 ('Grenada', 16.03, 8.08, 3.13),
 ('Cook Islands', 14.33, 8.03, None),
 ('Tanzania', 36.39, 8.0, 0.54),
 ('Palau', 11.05, 7.99, 0.71),
 ('Burma', 18.39, 7.96, 0.28),
 ('Dominica', 15.41, 7.91, 5.38),
 ('Haiti', 22.31, 7.83, 2.76),
 ('Thailand', 11.19, 7.8, 0.0),
 ('Djibouti', 23.65, 7.73, 6.06),
 ('Cuba', 9.9, 7.72, 3.66),
 ('Cambodia', 23.83, 7.68, 0.32),
 ('Jersey', 11.91, 7.68, 3.76),
 ('Sudan', 29.19, 7.66, 4.29),
 ('Laos', 24.25, 7.63, 1.09),
 ('Comoros', 27.84, 7.57, 2.53),
 ('China', 12.49, 7.53, 0.44),
 ('Eritrea', 30.0, 7.52, 0.0),
 ('Saint Helena, Ascension, and Tristan da Cunha', 9.88, 7.44, 